File: API/src/main/java/com/ltizzi/ecommerce/model/stock/StockEntity.java
package com.ltizzi.ecommerce.model.stock;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.ltizzi.ecommerce.model.product.ProductEntity;
import com.ltizzi.ecommerce.model.stockEntry.StockEntryEntity;
import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.SQLDelete;
import org.hibernate.annotations.UpdateTimestamp;
import org.hibernate.annotations.Where;

import java.sql.Timestamp;
import java.util.ArrayList;
import java.util.List;

/**
 * @author Leonardo Terlizzi
 */

@Entity
@Data
@NoArgsConstructor
@Table(name = "stocks")
@SQLDelete(sql = "UPDATE stocks SET soft_delete = true where stock_id=?")
@Where(clause = "soft_delete=false")
public class StockEntity {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long stock_id;

    @OneToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "product_id", nullable = false)
    @JsonIgnoreProperties({"hibernateLazyInitializer", "handler"})
    private ProductEntity product;

    private int cantidad;

    @OneToMany(fetch = FetchType.LAZY)
//    @JoinColumn(name = "entries", nullable = false)
    private List<StockEntryEntity> entries = new ArrayList<>();

    @CreationTimestamp
    @Column(updatable = false)
    private Timestamp createdAt;

    @UpdateTimestamp
    private Timestamp updatedAt;

    @Column(name = "soft_delete")
    private Boolean soft_delete = Boolean.FALSE;
}

File: Server/src/entities/StockEntity.ts
import {
  Column,
  CreateDateColumn,
  Entity,
  JoinColumn,
  OneToMany,
  OneToOne,
  PrimaryGeneratedColumn,
  UpdateDateColumn,
} from "typeorm";
import { Product } from "../models/Product";
import { Stock } from "../models/Stock";
import { StockEntry } from "../models/StockEntry";
import { ProductEntity } from "./ProductEntity";
import { StockEntryEntity } from "./StockEntryEntity";

@Entity("stocks")
export class StockEntity implements Stock {
  @PrimaryGeneratedColumn()
  stock_id!: number;

  @OneToOne(() => ProductEntity, { eager: true, nullable: false })
  @JoinColumn({ name: "product", referencedColumnName: "product_id" })
  product!: Product;

  @Column()
  product_id!: number;

  @Column()
  cantidad!: number;

  @OneToMany(() => StockEntryEntity, (entry) => entry.stock, {
    eager: true,
    nullable: false,
  })
  @JoinColumn({ name: "entry_id", referencedColumnName: "shop_entry_id" })
  entries!: StockEntry[];

  @CreateDateColumn({
    type: "timestamp",
    default: () => "CURRENT_TIMESTAMP",
  })
  createdAt!: Date;

  @UpdateDateColumn({
    type: "timestamp",
    default: () => "CURRENT_TIMESTAMP",
    onUpdate: "CURRENT_TIMESTAMP",
  })
  updatedAt!: Date;

  @Column({ default: false })
  soft_delete!: boolean;
}
